v3.0 - refactoring

JanWielgus opened the pull request.

Aleksy-niamik committed:
changed architecture of tasker - header files

Aleksy-niamik committed:
implemented new methods in Tasker class

JanWielgus commented:
Consider putting all classes in some namespace

JanWielgus committed:
Bug fixes

JanWielgus committed:
Updated Tasker example program

JanWielgus committed:
Examples update

JanWielgus committed:
Rebuilt frequency test program

Aleksy-niamik committed:
added briefs

Aleksy-niamik committed:
Merge branch 'refactoring' of https://github.com/JanWielgus/Tasker into refactoring

Aleksy-niamik committed:
added conditional compilation of load calculations

Aleksy-niamik committed:
fixed conditional compilation in getLoad

Aleksy-niamik committed:
added MinTaskInterval

JanWielgus committed:
Added Tasker example program

JanWielgus commented on TaskerConfig.h:
```diff
@@ -0,0 +1,16 @@
+#define TASKER_LOAD_FILTER_BETA 0.997f          // If load changes too rapidly or too slowly, adjust this value (bigger->change is slover)
+#define TASKER_LOAD_FILTER_BETA_COFACTOR 0.003f // this should be 1 - TASKER_LOAD_FILTER_BETA
+
+#endif
```
Leave a new line at the end of each file

JanWielgus commented on TaskerConfig.h:
```diff
@@ -0,0 +1,16 @@
+#ifndef TASKERCONFIG_H
+#define TASKERCONFIG_H
+
+#define TASKER_LOAD_CALCULATIONS 0              // 1 means "load calculations enabled", 0 means disabled "load calculations disabled"
```
I think that TASKER_LOAD_CALCULATIONS could be just a define. Then check #ifdef XXX instead of #if XXX == 1

JanWielgus commented on TaskerConfig.h:
```diff
@@ -0,0 +1,16 @@
+#define TASKERCONFIG_H
+
+#define TASKER_LOAD_CALCULATIONS 0              // 1 means "load calculations enabled", 0 means disabled "load calculations disabled"
+#define TASKER_LOAD_FILTER_BETA 0.997f          // If load changes too rapidly or too slowly, adjust this value (bigger->change is slover)
```
Add that value is: 0 < value < 1

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+        uint32_t nextExecutionTime_us;
+    };
+
+    const uint8_t MaxAmtOfTasks;    //  max amount of tasks - set once in the constructor
```
"- set once in the constructor" this info is useless here.
In classes all consts have to be set immediately after name or in the ctor.

Also this name can be changed to MaxTasksAmt

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    };
+
+    const uint8_t MaxAmtOfTasks;    //  max amount of tasks - set once in the constructor
+    Task* tasks;                    // static array of MaxAmtOfTasks size
```
This array is not static in case of c++. It is allocated dynamically. Just it's size remains unchanged.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+
+    const uint8_t MaxAmtOfTasks;    //  max amount of tasks - set once in the constructor
+    Task* tasks;                    // static array of MaxAmtOfTasks size
+    uint8_t amountOfTasks = 0;      // current amount of tasks (at most MaxAmtOfTasks)
```
Better name is tasksAmt or tasksAmount

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    const uint8_t MaxAmtOfTasks;    //  max amount of tasks - set once in the constructor
+    Task* tasks;                    // static array of MaxAmtOfTasks size
+    uint8_t amountOfTasks = 0;      // current amount of tasks (at most MaxAmtOfTasks)
+    uint32_t currentTime = 0;       // current time (used in loop() method)
```
In my opinion name of this variable should be something like: lastMicrosTime

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    ~Tasker();
+
+    /**
+     * @brief Add task with it's frequency.
```
More like: "Add task and set it's execution frequency" (same in addTask_us() brief but explaining what is that interval)

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+     * @param task Pointer to concrete class that implements IExecutable interface.
+     * @param frequency Task running frequency (in Hz).
+     * @return true if the task was successfully added
+     * @return false otherwise
```
There should be only one @return in a brief. For example add comma after the first line and remove @return in the second line (in effect all new lines will be changed to spaces).

The same everywhere.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    /**
+     * @brief Add task with it's interval.
+     * @param task Pointer to concrete class that implements IExecutable interface.
+     * @param interval_us Task running interval (in us).
```
Maybe explain a little bit what in practice interval is

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * @brief Set task's frequency (in Hz) 
+     * @param task Task which fequency is to change
```
Maybe better is: Previously added task which frequency you want to change.
(typo in frequency)

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+     * @brief Set task's frequency (in Hz) 
+     * @param task Task which fequency is to change
+     * @param frequency_Hz New frequency (in Hz)
+     * @return true if the task was found and therefore its frequency was changed
```
I would write shorter that return false if task was not found, otherwise true.
Same in line 95, 104 and other places if I missed something.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+
+    /**
+     * @brief Set task's interval (in us)
+     * @param task Task which interval is to change
```
Maybe better is: Previously added task which interval you want to change.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    /**
+     * @brief Pause task for a specified time (in us)
+     * @param task Task to pause
+     * @param pauseTime_us Time for task to be stopped (in us)
```
"paused" instead of "stopped"

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    bool pauseTask_s(IExecutable* task, float pauseTime_s);
+
+    /**
+     * @param task Task whose frequency (in Hz) is returned
```
I would write: Previously added task whose frequency you want to check.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    float getTaskFrequency_Hz(IExecutable* task);
+
+    /**
+     * @param task Task whose interval (in us) is returned
```
I would write: Previously added task whose interval you want to check.
(Similar in version with seconds)

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+    uint32_t getCurrentTime_micros();
+
+    /**
+     * @brief This should be the only method in arduino's loop(). Execute one Task execution time checking.
```
The last sentence is confusing.

JanWielgus commented on Tasker.h:
```diff
@@ -0,0 +1,155 @@
+
+
+private:
+    void calculateNextTask();
```
I think there should be a short brief

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+/**
+ * @file SimpleTasker.cpp
+ * @author Jan Wielgus & Aleksy Walczak
+ * @brief 
```
Remove that line

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+
+bool Tasker::addTask_Hz(IExecutable* task, float frequency_Hz)
```
This should be more simple and do the job as well:

```cpp
bool Tasker::addTask_Hz(IExecutable* task, float frequency_Hz)
{
    if (amountOfTasks >= MaxAmtOfTasks || task == nullptr || frequency_Hz <= 0)
        return false;

    uint32_t newInterval_us = 1000000.0 / frequency_Hz + 0.5f;
    return addTask_us(task, newInterval_us);
}
```

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+
+void Tasker::calculateNextTask()
```
This method is the last one in the header file, so put it at the end of the cpp file. Check order of all methods.

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+
+bool Tasker::setTaskFrequency(IExecutable* task, float frequency_Hz)
```
I would use just setTaskInterval_us() method inside to avoid code duplication.

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+
+bool Tasker::setTaskFrequency(IExecutable* task, float frequency_Hz)
+{
+    for (uint8_t i = 0; i < amountOfTasks-1; ++i)
```
without -1 (but irrelevant if method body will be replaced with setTaskInterval_us() method call)

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+}
+
+
+bool Tasker::pauseTask_s(IExecutable* task, float pauseTime_s)
```
I think better will be to use pauseTask_us() method inside

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+
+bool Tasker::setTaskInterval_us(IExecutable* task, uint32_t interval_us)
+{
+    for (uint8_t i = 0; i < amountOfTasks; ++i)
+    {
+        if (tasks[i].executable == task)
```
A thing to consider is to move finding a task to different method. This functionality repeats over and over again, so such method could for example return boolean if found a task in list of tasks added to the Tasker.

JanWielgus commented on Tasker.cpp:
```diff
@@ -0,0 +1,258 @@
+    return (uint32_t)-1;
+}
+
+float Tasker::getTaskInterval_s(IExecutable* task) 
```
This method also could just use getTaskInterval_us() method inside.

JanWielgus commented on IExecutable.h:
```diff
@@ -15,6 +15,9 @@ class IExecutable
     virtual ~IExecutable() {}
 
+    /**
+     * @brief Do some action - used in Tasker loop to be executed regularly with some frequency
```
I would just leave "Do some action"

JanWielgus committed:
Small code formatting change

JanWielgus commented:
Features to test:
- [x] adding tasks
- [x] removing tasks
- [x] changing task parameters
- [x] pausing a task
- [x] task params getters
- [x] tasks amount getter
- [x] tasker load feature
- [ ] sleeping function's energy saving

JanWielgus committed:
Bug fix after tests

JanWielgus committed:
Small tasker example layout upgrade

JanWielgus committed:
Moved examples to the separate folder

JanWielgus committed:
Merge branch 'refactoring' of https://github.com/JanWielgus/Tasker into refactoring

Aleksy-niamik committed:
applied briefs corrections and small methods refactoring

Aleksy-niamik committed:
made loop method inline

JanWielgus committed:
Added waiting function implementation ready for testing

JanWielgus commented:
(To myself) Remember to update example programs in other libraries that uses Tasker

JanWielgus committed:
Lot of refactoring

Aleksy-niamik committed:
changed name of getExecStartTime to getLastTaskExecutiionTime

Aleksy-niamik committed:
Merge pull request #18 from JanWielgus/custom-waiting-function-feature

Sleeping function feature and small refactoring of refactoring

Aleksy-niamik committed:
added processor overload callback and made sleeping function feature conditional

Aleksy-niamik commented:
> Features to test:
> 
> * [x]  adding tasks
> * [x]  removing tasks
> * [x]  changing task parameters
> * [x]  pausing a task
> * [x]  task params getters
> * [x]  tasks amount getter
> * [ ]  tasker load feature
> * [ ]  sleeping function's energy saving

JanWielgus committed:
Mainly brief and variable names updates

JanWielgus committed:
Small corrections

JanWielgus committed:
Changed sleepingFunction to sleepFunction

JanWielgus committed:
Examples refactoring

JanWielgus committed:
Some development

- removed processor overloaded callback
- rebuild loop() method
- some refactoring

JanWielgus committed:
Small refactoring of example program

JanWielgus committed:
Macros indentation corrections

Aleksy-niamik reviewed (approved):
Dobrze jes
